### ISSUE TITLE
Constructor assigns wrong assay values to wrong samples and features

### ISSUE THREAD

**Issue by TuomasBorman** [Bioconductor/SummarizedExperiment#56]
Hi!

I found out a possible bug from SE constructor. 

`assay` data's sample and feature names are ordered based on `colData` and `rowData`, respectively. However, values are not ordered correspondingly --> **wrong values goes to wrong samples and features**

Here is an example that shows the behavior:
```
library(SummarizedExperiment)

# assay data
counts <- rbind(rep(0, 3), matrix(1:9, nrow = 3))
colnames(counts) <- paste0("sample", 1:3)
rownames(counts) <- paste0("OTU", seq_len(4))
# row data
rowData <- data.frame(Species = rep(c("S1", "S2"), each = 2),
                      row.names = rownames(counts),
                      stringsAsFactors = FALSE)
# col data
colData <- DataFrame(Treatment=c("X", "X", "Y"),
                     row.names=colnames(counts))

# Modify the order of samples in assay data
counts_sample_mod <- counts[, c("sample2", "sample3", "sample1")]
# Modify the order of features in assay data
counts_feat_mod <- counts[c( "OTU2", "OTU4", "OTU1", "OTU3"), ]

# Create three different SE objects, which have differently ordered assay data
se <- SummarizedExperiment(assays=SimpleList(counts=counts),
                           colData=colData,
                           rowData = rowData)
se_sample_mod <- SummarizedExperiment(assays=SimpleList(counts=counts_sample_mod),
                                      colData=colData,
                                      rowData = rowData)
se_feat_mod <- SummarizedExperiment(assays=SimpleList(counts=counts_feat_mod),
                                    colData=colData,
                                    rowData = rowData)

# Sample and feature names are ordered based on colData and rowData. 
# However, values are not ordered correspondingly.
# Original
counts
assay(se, "counts")
# Samples modified
counts_sample_mod
assay(se_sample_mod, "counts")
# Features modified
counts_feat_mod
assay(se_feat_mod, "counts")
```

Outputs of those assay tables:
```
> # Original
> counts
     sample1 sample2 sample3
OTU1       0       0       0
OTU2       1       4       7
OTU3       2       5       8
OTU4       3       6       9
> assay(se, "counts")
     sample1 sample2 sample3
OTU1       0       0       0
OTU2       1       4       7
OTU3       2       5       8
OTU4       3       6       9
> # Samples modified
> counts_sample_mod
     sample2 sample3 sample1
OTU1       0       0       0
OTU2       4       7       1
OTU3       5       8       2
OTU4       6       9       3
> assay(se_sample_mod, "counts")
     sample1 sample2 sample3
OTU1       0       0       0
OTU2       4       7       1
OTU3       5       8       2
OTU4       6       9       3
> # Features modified
> counts_feat_mod
     sample1 sample2 sample3
OTU2       1       4       7
OTU4       3       6       9
OTU1       0       0       0
OTU3       2       5       8
> assay(se_feat_mod, "counts")
     sample1 sample2 sample3
OTU1       1       4       7
OTU2       3       6       9
OTU3       0       0       0
OTU4       2       5       8
```

BR,
Tuomas Borman

**Comment by TuomasBorman** [comment]
```
> sessionInfo()
R version 4.1.0 (2021-05-18)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19042)

Matrix products: default

locale:
[1] LC_COLLATE=English_Finland.1252  LC_CTYPE=English_Finland.1252    LC_MONETARY=English_Finland.1252 LC_NUMERIC=C                    
[5] LC_TIME=English_Finland.1252    

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] SummarizedExperiment_1.22.0 Biobase_2.52.0              GenomicRanges_1.44.0        GenomeInfoDb_1.28.0         IRanges_2.26.0             
[6] MatrixGenerics_1.4.0        matrixStats_0.59.0          S4Vectors_0.30.0            BiocGenerics_0.38.0        

loaded via a namespace (and not attached):
 [1] lattice_0.20-44        bitops_1.0-7           grid_4.1.0             zlibbioc_1.38.0        XVector_0.32.0         Matrix_1.3-4          
 [7] tools_4.1.0            RCurl_1.98-1.3         DelayedArray_0.18.0    compiler_4.1.0         GenomeInfoDbData_1.2.6
```

**Comment by LiNk-NY** [comment]
Hi Tuomas, @TuomasBorman 

Thank you for providing a detailed report. 
It appears that the `colData` and `rowData` orderings take precedence over the ordering of the `assay`.
Perhaps it should be the other way around. 
We will investigate and get back to you. 

Best regards,
Marcel

**Comment by hpages** [comment]
@TuomasBorman, @LiNk-NY,

I can think of 2 ways to deal with this:
1. The rows/cols of the supplied assay (or assays) gets reordered to align with the rownames/colnames of the SE object (these are inferred from the rownames of the rowData/colData).
2. We raise an error if one of the supplied assays has rownames and/or colnames that don't match those of the SE object.

The second option (2.) is consistent with the `assay()`/`assays()` setters so that's my preference. The first option (1.) is more complicated than it sounds e.g. it's not clear what we should do if the rownames/colnames of the user-supplied array are completely different from those of the SE object or if only a subset of them match those of the SE object.

I've a patch ready for (2.) that I'm going to commit today. It might break a few packages on the [devel builds](https://bioconductor.org/checkResults/3.14/bioc-LATEST/), we'll see. Hopefully it won't be too bad. In any case they should be easy to fix.

H.

**Comment by TuomasBorman** [comment]
@hpages 

Sounds good. In my opinion, option 1 would be a nice feature for some users, I guess. However, in addition of being simpler, option 2 fits better to the nature of SE of being just a data container, so no too fancy things should be going on

-Tuomas

**Comment by hpages** [comment]
Done in **SummarizedExperiment** 1.23.2 (see commit 1e1da5b79a035a0f597dedcba42d2cec7a2406c3).

I've manually run `R CMD check` on a few selected reverse deps: **SingleCellExperiment**, **TreeSummarizedExperiment**, **zinbwave**, **VariantAnnotation**, **xcms**, **iSEE**, **velociraptor**, and **tximeta**. All went fine except for a failed test in **SingleCellExperiment** that I will fix ASAP.

Now **SummarizedExperiment** has 422 direct reverse deps so I will let the daily builds run all the checks for me. Fingers crossed that we don't see a blood bath on Monday's report (I committed my change too late for Saturday's report and we don't have reports on Sundays).

**Comment by hpages** [comment]
**SingleCellExperiment** fixed (commit b96ffaff, version 1.15.2)

@Marcel Hey Marcel, this change to the `SummarizedExperiment()` constructor function causes some of the unit tests in **MultiAssayExperiment** to fail. Sorry for that. Do you think you can take a look? The easy fix is to remove the offending rownames and/or colnames from the supplied assays (naked assays are always accepted). Thanks!

**Comment by hpages** [comment]
Other randomly picked up reverse deps that went thru a manual `R CMD check` on my laptop: **AllelicImbalance**, **batchelor**, **BiocSklearn**, **bsseq**, **CATALYST**, **ChIPpeakAnno**, **TCGAutils**, and **zellkonverter**. They all passed successfully.
